### PR TITLE
Build SU2 and run regressions on self hosted aarch64 runner scheduled

### PR DESCRIPTION
**Pull request by martin-g**
## Proposed Changes

Add `workflow_call` event to `regression.yml` Github Actions workflow. This way it is possible to [re-use](https://docs.github.com/en/actions/using-workflows/reusing-workflows) it from another workflow.

Add a new workflow - `scheduled-arm64.yml`. It is a scheduled workflow that just executes `regressions.yml` on a self-hosted runner.

## Related Work

https://github.com/su2code/Docker-Builds/pull/1 - A PR that updates the build and test Docker images to Ubuntu 20.04 so that they could be used on both AMD64 and ARM64.

## PR Checklist

- [X] I am submitting my contribution to the develop branch.
- [X] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags, or simply --warnlevel=2 when using meson).
- [ ] My contribution is commented and consistent with SU2 style.
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

### CODE REVIEW

File: .github/workflows/regression.yml [hunk 1]
```diff
@@ -9,14 +9,20 @@ on:
     branches: 
       - 'develop'
       - 'master'
+  workflow_call:
+    # workflows: ["Regression and Unit Post-Validation on Linux ARM64"]
+    # types: [requested]
+    # branches:
+      # - 'develop'
```
Comment (martin-g) [inline .github/workflows/regression.yml:16]: The lines will be uncommented later to lock the caller only to `scheduled-arm64.yml`

File: .github/workflows/regression.yml [hunk 2]
```diff
@@ -9,14 +9,20 @@ on:
...
 
 jobs:
   build:
     name: Build SU2
     strategy:
       fail-fast: false
       matrix: 
-        config_set: [BaseMPI, ReverseMPI, ForwardMPI, BaseNoMPI, ReverseNoMPI, ForwardNoMPI, BaseOMP, ReverseOMP, ForwardOMP]
+        #config_set: [BaseMPI, ReverseMPI, ForwardMPI, BaseNoMPI, ReverseNoMPI, ForwardNoMPI, BaseOMP, ReverseOMP, ForwardOMP]
+        config_set: [BaseMPI]
```
Comment (martin-g) [inline .github/workflows/regression.yml:25]: Temporarily reduced the config_set to reduce the build time. Once it works I will re-enable the complete list again.

File: .github/workflows/regression.yml [hunk 3]
```diff
@@ -45,7 +51,7 @@ jobs:
           key: ${{ matrix.config_set }}-${{ github.sha }}
           restore-keys: ${{ matrix.config_set }}
       - name: Build 
-        uses: docker://su2code/build-su2:20191105
+        uses: docker://ghcr.io/martin-g/su2code/build-su2:2
```
Comment (martin-g) [inline .github/workflows/regression.yml:54]: Related to https://github.com/su2code/Docker-Builds/pull/1. I need to use the Docker image on the self-hosted Linux ARM64 runner and this requires updating the base image to Ubuntu 20.04 that is multi-platform.

The url to the Docker image will be updated to an official one once https://github.com/su2code/Docker-Builds/pull/1 is merged.

File: .github/workflows/regression.yml [hunk 4]
```diff
@@ -89,7 +95,7 @@ jobs:
           if [ -d "${{format('Forward{0}', matrix.tag)}}" ]; then cp -r ${{format('Forward{0}', matrix.tag)}}/. install/bin/; fi
           chmod a+x install/bin/*
       - name: Run Tests in Container
-        uses: docker://su2code/test-su2:20200303
+        uses: docker://ghcr.io/martin-g/su2code/test-su2:1
```
Comment (martin-g) [inline .github/workflows/regression.yml:98]: Same as with the `build-su2` image.

File: .github/workflows/scheduled-arm64.yml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+name: Post-Validation on Linux ARM64
+
+on:
+  schedule:
+    - cron: "*/2 * * * *"
```
Comment (martin-g) [inline .github/workflows/scheduled-arm64.yml:5]: Every 2 minutes to be able to see the result quickly. Once everything works the schedule will be once per day or something similar.

**Comment by martin-g** [comment]
Initially I tried to work on this task in my personal fork - https://github.com/martin-g/SU2/pull/1.
The problem there was that `compileSU2.sh` clones https://github.com/su2code/SU2 and tries to find the PR ref in it and fails.
Now this problem should be gone but the expected issue is that the self-hosted runner won't be found because it is not yet registered for this Github repository. 
@pcarruscag Please send me the preferred hardware specs (CPU cores, RAM, disk, ...) at martin.grigorov at gmail dot com and I will reserve a VM!

**Comment by martin-g** [comment]
@pcarruscag Could you please enable the Github Actions checks for this PR ?
They are disabled because it is my first PR in this repository.
Thank you!

**Comment by martin-g** [comment]
The Github Actions checks passed after making two changes:

1) https://github.com/su2code/SU2/pull/1619/commits/efe98fe6070a0cb51f1082a9599363786e4d65ea
I needed to lower the warnlevel due to problems in `externals/cgns/hdf5`.
Maybe the HDF5 files should be updated to a newer version ?! But it is not clear to me which version has been used in https://github.com/su2code/SU2/pull/1500. 
@MicK7 Do you have an idea how to fix th

2) https://github.com/su2code/SU2/pull/1619/commits/26140223e5838a6856c0b3c02163a826256306b1
A workaround recommended at https://github.com/su2code/SU2/issues/1568#issuecomment-1083104460

3) The regression tests [fail](https://github.com/su2code/SU2/runs/6387003184?check_suite_focus=true) because the new `test-su2` Docker image is `private`. I will need someone with higher privileges than me to make it public.

**Comment by martin-g** [comment]
@pcarruscag The build passes now after updating the regression tests expectations!

Should I continue working on this PR for the ARM64 part or should we merge it (without the new `regressions-arm64.yml`) and do the rest in another PR ?

Also what is your opinion on the change for the `warnlevel` ? Without it it is not possible to compile HDF5

**Comment by martin-g** [comment]
The regression tests fail on Ubuntu 

```
==================================================================
Summary of the parallel tests
python version: 3.8.10 (default, Mar 15 2022, 12:22:08) 
[GCC 9.4.0]
  passed - discadj_naca0012
  passed - discadj_cylinder3D
  passed - discadj_arina2k
  passed - ea_naca64206
  passed - discadj_rans_naca0012_sa
  passed - discadj_rans_naca0012_sst
  passed - discadj_incomp_NACA0012
  passed - discadj_incomp_cylinder
  passed - discadj_incomp_turb_NACA0012_sa
  passed - discadj_incomp_turb_NACA0012_sst
  passed - discadj_axisymmetric_rans
  passed - unsteady_cylinder
  passed - unsteady_cylinder_windowed_average_AD
* FAILED - unsteady_cylinder_windowed_average
  passed - unsteady_cylinder_DT_1ST
  passed - pitchingNACA0012
  passed - transonic_stator
* FAILED - discadj_fea
  passed - discadj_heat
  passed - discadj_fsi
* FAILED - discadj_fsi_airfoil
  passed - discadj_cht
  passed - da_sp_pinArray_cht_2d_dp_hf
  passed - da_sp_pinArray_cht_2d_mf
  passed - da_unsteadyCHT_cylinder
  passed - discadj_topol_optim
  passed - unsteady_NACA0012_restart_adjoint
  passed - unsteady_NACA0012_adjoint_only
  passed - restart_shape_optimization
* FAILED - dyn_discadj_fsi
* FAILED - grad_smooth_sob
Tests failed
```

It is the same version of Python and GCC as on the Github Actions hosted runner.
We will need logic to differentiate the expected results on x86_64 and aarch64 :-/

**Comment by martin-g** [comment]
@pcarruscag I've reverted the `warnlevel` back to `=3` and added some exclusions for errors and warnings in `externals/**`!

But there are also some problems in SU2's code (we should see them as build errors in the new checks).

With 
```
diff --git a/meson.build b/meson.build
index 4c95a97047..1a34b8ac1d 100644
--- a/meson.build
+++ b/meson.build
@@ -22,6 +22,12 @@ if build_machine.system() != 'windows'
     default_warning_flags += ['-Wno-empty-body']
   endif
   default_warning_flags += ['-Wno-unused-parameter',
+    '-Wno-error=cast-function-type',
+    '-Wno-cast-function-type',
+    '-Wno-error=stringop-truncation',
+    '-Wno-stringop-truncation',
+    '-Wno-error=missing-field-initializers',
+    '-Wno-missing-field-initializers',
                             '-Wno-deprecated-declarations']
```
the build passes locally but I guess those should be fixed instead of suppressed.

**Comment by martin-g** [comment]
Here is one such error - https://github.com/su2code/SU2/runs/6524060568?check_suite_focus=true
```
/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/op_inln.h:123:59: error: cast between incompatible function types from ‘void (*)(const void*, void*, int, const MPI::Datatype&)’ to ‘void (*)(void*, void*, int*, ompi_datatype_t**)’ [-Werror=cast-function-type]
  123 |     ompi_op_set_cxx_callback(mpi_op, (MPI_User_function*) func);
      |                                                           ^~~~
cc1plus: all warnings being treated as errors
[524/816] Compiling C++ object Common/src/libSU2Common.a.p/linear_algebra_blas_structure.cpp.o
FAILED: Common/src/libSU2Common.a.p/linear_algebra_blas_structure.cpp.o 
ccache c++ -ICommon/src/libSU2Common.a.p -ICommon/src -I../Common/src -I../externals/CLI11 -I../externals/cgns/adf -I../externals/cgns/adfh -Iexternals/cgns -I../externals/cgns -Iexternals/cgns/hdf5 -I../externals/cgns/hdf5 -I../externals/metis/include -I../externals/metis/GKlib -I../externals/metis/libmetis -I../externals/parmetis/include -I../externals/parmetis/libparmetis -Iexternals/tecio -I../externals/tecio -Iexternals/tecio/teciompisrc -I../externals/tecio/teciompisrc -I../externals/mel -I/usr/lib/x86_64-linux-gnu/openmpi/include -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -Wextra -Wpedantic -Werror -std=c++11 -O3 -fPIC -pthread -Wno-empty-body -Wno-unused-parameter -Wno-deprecated-declarations -DHAVE_CGNS -DNDEBUG -DHAVE_MPI -DHAVE_METIS -DHAVE_PARMETIS -DHAVE_TECIO -DTECIOMPI -MD -MQ Common/src/libSU2Common.a.p/linear_algebra_blas_structure.cpp.o -MF Common/src/libSU2Common.a.p/linear_algebra_blas_structure.cpp.o.d -o Common/src/libSU2Common.a.p/linear_algebra_blas_structure.cpp.o -c ../Common/src/linear_algebra/blas_structure.cpp
In file included from /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/mpicxx.h:277,
                 from /usr/lib/x86_64-linux-gnu/openmpi/include/mpi.h:2868,
                 from ../Common/src/linear_algebra/../../include/parallelization/mpi_structure.hpp:33,
                 from ../Common/src/linear_algebra/../../include/CConfig.hpp:31,
                 from ../Common/src/linear_algebra/blas_structure.cpp:29:
/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/op_inln.h: In member function ‘virtual void MPI::Op::Init(void (*)(const void*, void*, int, const MPI::Datatype&), bool)’:
/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/op_inln.h:121:46: error: cast between incompatible function types from ‘void (*)(void*, void*, int*, ompi_datatype_t**, void (*)(void*, void*, int*, ompi_datatype_t**))’ to ‘void (*)(void*, void*, int*, ompi_datatype_t**)’ [-Werror=cast-function-type]
  121 |     (void)MPI_Op_create((MPI_User_function*) ompi_mpi_cxx_op_intercept,
      |                                              ^~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/op_inln.h:123:59: error: cast between incompatible function types from ‘void (*)(const void*, void*, int, const MPI::Datatype&)’ to ‘void (*)(void*, void*, int*, ompi_datatype_t**)’ [-Werror=cast-function-type]
  123 |     ompi_op_set_cxx_callback(mpi_op, (MPI_User_function*) func);
      |                                                           ^~~~
cc1plus: all warnings being treated as errors
[525/816] Compiling C++ object externals/tecio/teciompisrc/libteciompi.a.p/TecioSZL.cpp.o
ninja: build stopped: subcommand failed.
```

**Comment by pcarruscag** [comment]
Thank you for looking into that.
We can suppress new warnings for now while you continue the ARM work and fix in a different PR. If the warnings come from third parties (MPI in your example) it is difficult to fix them.

File: meson.build [hunk 6]
```diff
@@ -22,7 +22,16 @@ if build_machine.system() != 'windows'
...
-                            '-Wno-deprecated-declarations']
+                            '-Wno-deprecated-declarations',
+                            '-Wno-error=cast-function-type',
+                            '-Wno-cast-function-type',
+                            '-Wno-error=stringop-truncation',
+                            '-Wno-stringop-truncation',
+                            '-Wno-error=missing-field-initializers',
+                            '-Wno-missing-field-initializers',
+                            '-Wno-error=class-memaccess',
+                            '-Wno-class-memaccess'
```
Comment (martin-g) [inline meson.build:33]: Some of those exclusions are for "external" code like openmpi but some of them are for internal/SU2 code and code be removed after fixing the issues.

**Comment by martin-g** [comment]
@pcarruscag The aarch64 related changes are almost done!

There are few tests which fail with this error, e.g. serial_regression.py:
```
==================== Start Test: turb_naca0012_3c ====================
/github/workspace/tests/TestCases/rans_uq/naca0012
turb_naca0012_3c: FAILED
Output for the failed case
execution command: SU2_CFD turb_NACA0012_uq_3c.cfg.autotest > turb_NACA0012_uq_3c.cfg.log 2>&1
ERROR: The code was not able to get to the "Begin solver" section.
ERROR: The iteration number 10 could not be found.
CPU architecture=aarch64
test_iter=10
test_vals (stored): -5.584300, 0.931352, 0.248990, -0.109405
sim_vals (computed): 
delta_vals: 
test duration: 0.00 min
==================== End Test: turb_naca0012_3c ====================



Error in "void CConfig::SetConfig_Parsing(std::istream&)": 
-------------------------------------------------------------------------
Line 26 SST_OPTIONS: invalid option name. Check current SU2 options in config_template.cfg.
Did you mean FFD_ITERATIONS?

------------------------------ Error Exit -------------------------------
```

Any idea what could be the reason this to fail only on ARM64 ?

**Comment by martin-g** [comment]
About the new CodeFactor issues: I aligned the x86_64 and aarch64 testVals so it is easier (for the naked eye) to compare the differences. But now CodeFactor complains about the extra spaces before the `=`. Should I remove the extra spaces or this could be configured to silent it somewhere ?

**Comment by martin-g** [comment]
@pcarruscag The PR is ready for review!
There are two regression tests which always fail - https://github.com/su2code/SU2/runs/7019382870?check_suite_focus=true
`polar_naca0012` in `serial_regression.py` and `parallel_regression.py`. The computed values are different in each run :-/

**Comment by martin-g** [comment]
Since I used the Github UI to merge the latest updates in `develop` branch now it is not possible to rebase+squash the commits in this branch.
Unless you have a better idea I plan to just copy the contents of the modified files in this PR to a new branch and open a new PR.

**Comment by pcarruscag** [comment]
Usually we dont squash-merge code PRs, so it's fine.

**Review by pcarruscag — APPROVED** [review]

**Pull request merged**